torchtune/modules/tokenizers/_utils.py

```diff
@@ -4,58 +4,178 @@
 # This source code is licensed under the BSD-style license found in the
 # LICENSE file in the root directory of this source tree.
 
-from typing import Iterator, List, Protocol, Set
+import json
+from typing import Dict, List, Optional, Protocol, Tuple
 
 from torchtune.data._types import Message
+from torchtune.data._utils import truncate
 
 
-class Tokenizer(Protocol):
-    """Abstract tokenizer"""
-
-    bos_id: int
-    eos_id: int
-    pad_id: int
-    stop_tokens: Set[int]  # Tokens indicating that generation should stop
+class BaseTokenizer(Protocol):
+    """
+    Abstract token encoding model that implements ``encode`` and ``decode`` methods.
+    """
 
     def encode(self, text: str, **kwargs) -> List[int]:
         """
-        Given a string, return the a list of token ids.
+        Given a string, return the encoded list of token ids.
+
+        Args:
+            text (str): The text to encode.
+
+        Returns:
+            List[int]: The encoded list of token ids.
         """
+        pass
 
     def decode(
-        self, token_ids: List[int], add_bos: bool, add_eos: bool, **kwargs
+        self, token_ids: List[int], include_special: bool = False, **kwargs
```

Comment:
Maybe I'm being dense, I don't see `include_special` being used anywhere?

Reply:
It's not. I may just remove it. It's not straightforward to enable including special tokens for SentencePiece, at least.